book/visual-effects.md

```diff
@@ -989,316 +996,169 @@ def parse_blend_mode(blend_mode_str):
+like putting a second piece of paper (called a *mask*) over the first
+one, and then using scissors to cut along the mask's edge.
+
+There are all sorts of powerful methods[^like-clip-path] for clipping
+content on the web, but the most common form involves the `overflow`
+property. This property has lots of possible values,[^like-scroll] but
+let's focus here on `overflow: clip`, which cuts off contents of an
+element that are outside the element's bounds.
+
+[^like-clip-path]: The CSS [`clip-path` property][mdn-clip-path] lets
+specify a mask shape using a curve, while the [`mask`
+property][mdn-mask] lets you instead specify a image URL for the mask.
+
+[^like-scroll]: For example, `overflow: scroll` adds scroll bars and
+    makes an element scrollable, while `overflow: hidden` is similar
+    but subtly different from `overflow: clip`.
+
+[mdn-mask]: https://developer.mozilla.org/en-US/docs/Web/CSS/mask
+
+[mdn-clip-path]: https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path
+
+Usually, `overflow: clip` is used with properties like `height` or
+`rotate` which can make an element's children poke outside their
+parent. Our browser doesn't support these, but there is one edge case
+where `overflow: clip` is relevant: with rounded corners. Consider
+this example:
+
+``` {.html.example}
+<div style="border-radius:30px;background-color:lightblue;overflow:clip">
+    This test text exists here to ensure that the "div" element is
+    large enough that the border radius is obvious.
 </div>
-
-Implementing circular clips is once again pretty easy with Skia in our back
-pocket, but has some differences from other visual effects. We'll make a new
-surface (the mask), draw the circle into it, and then blend it with
-destination-in compositing. Let's first get the code in place and then
-investigate more about how it works.
-
-Parse the `clip-path` CSS property:
-
-``` {.python}
-def parse_clip_path(clip_path_str):
-    if clip_path_str.find("circle") != 0:
-        return None
-    return int(clip_path_str[7:][:-2])
 ```
 
-And define how to paint it:
+That HTML looks like this:
 
-``` {.python}
-def center_point(rect):
-    return (rect.left() + (rect.right() - rect.left()) / 2,
-        rect.top() + (rect.bottom() - rect.top()) / 2)
-# ...
-
-def paint_clip_path(node, display_list, rect):
-    clip_path = node.style.get("clip-path")
-    if clip_path:
-        percent = parse_clip_path(clip_path)
-        if percent:
-            width = rect.right() - rect.left()
-            height = rect.bottom() - rect.top()
-            reference_val = \
-                math.sqrt(width * width +
-                    height * height) / math.sqrt(2)
-            radius = reference_val * percent / 100
-            (center_x, center_y) = center_point(rect)
-            display_list.append(CircleMask(
-                center_x, center_y, radius, rect))
-```
+<div style="border-radius:30px;background-color:lightblue;overflow:clip">
```

Comment:
We can add overflow:hidden as a hack. Also, I happen to know that the implementation is complete and it'll ship early next year...

Reply:
I'll resolve this later.